### DISCUSSION TITLE
Wrong sequence of files in montage

### DISCUSSION THREAD

**Discussion by ferropasha** [ImageMagick/ImageMagick#6900] [Help]
### ImageMagick version

7.0.9-1

### Operating system, version and so on

Windows 10

### Description

I stumbled upon a rather maddening issue I can't resolve on my own. I am experimenting with montages of files. My autoit script takes files in batches of eight and creates montages in 4x2 arrangements also labelling each one, autoit part here simply acts as a wrapper over IM to make a cmd line I need and runs it. The script looks like this
`magick montage $montage_list -tile 4x2 -geometry +10+20 -frame 10 $output_file` 
where $montage_list is autoit string variable that contains pairs of labels and file-paths to feed IM for montage (8 of them) and looks like this
`-label "some label for my file1" "path to my file1" -label "some label for my file2"' "path to my file2" `
The problem is output file I get has my files in wrong sequence. The first label-file pair gets **shuffled back to eigth place instead of opening the sequence**. And this behavior persists even if I change montage arrangement to say 3x2, or 4x or x2 or even discard -tile parameter entirely.
Then to eliminate any influence from autoit I temporarily discarded my wrapper script and simply run the command I need in plain cmd. **The result is the same, so it is definitely some quirck in IM behavior.** All files have names like 0001.tif, 0002.tif etc. so there can't be any issue with sorting.   
What is it then?
P.S. There appears to be some unknown problem with the files themselves because I converted them from tifs to jpegs with the same filenames and the command worked fine while with tifs there was a warning "Unknown field with tag 292 (0x124) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/1029. Is it possible that some tif metadata is responsible?

**Comment by snibgo** [comment]
See https://imagemagick.org/script/command-line-options.php#label

I suspect your TIFF files already have labels, and when each file is read, that will override your "-label" setting. Use "`magick yourtiff.tiff -verbose info:`" to check for this.

If that is the problem, then I suggest using `-set label "blah"` which will set the label for *all* images in the current list. You want to set the label for each image individually, so put them in parentheses.

Set `montage_list` to:
```
( "path to my file1" -set label "some label for my file1" ) ( "path to my file2" -set label "some label for my file2" )
```

**Reply by ferropasha** [reply]
Thank you for your reply but no luck so far. I rewrote my test command as you suggested so now it looks like this
`magick montage ( "h:\test\1.tif" -set label "1" ) ( "h:\test\2.tif" -set label "2" ) ( "h:\test\3.tif" -set label "3" ) ( "h:\test\4.tif" -set label "4" ) ( "h:\test\5.tif" -set label "5" ) ( "h:\test\6.tif" -set label "6" ) -tile 3x2 -geometry +10+20 -frame 10 -pointsize 60 -title "test"  test_montage.jpg`
Result is the same. In the montage  image 1.tif with label 1 is put in the last position as if the montage is supposed to be read from bottom-right corner not the top-left. The rest are in proper order starting from image 2.tif labelled 2 in the top-left corner, then 3 etc.  I also checked 1.tif with -verbose. There doesn't appear to be any label properties in the output (see below) 
```
 Image:
  Filename: 1.tif
  Format: TIFF (Tagged Image File Format)
  Mime type: image/tiff
  Class: DirectClass
  Geometry: 3283x4714+0+0
  Resolution: 300x300
  Print size: 10.9433x15.7133
  Units: PixelsPerInch
  Colorspace: sRGB
  Type: TrueColor
  Endianness: Undefined
  Depth: 8-bit
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
  Channel statistics:
    Pixels: 15476062
    Red:
      min: 1  (0.00392157)
      max: 255 (1)
      mean: 166.85 (0.654313)
      standard deviation: 33.7688 (0.132427)
      kurtosis: 2.30162
      skewness: -1.80607
      entropy: 0.787823
    Green:
      min: 6  (0.0235294)
      max: 255 (1)
      mean: 146.423 (0.57421)
      standard deviation: 30.2383 (0.118582)
      kurtosis: 1.84801
      skewness: -1.5997
      entropy: 0.807418
    Blue:
      min: 0  (0)
      max: 255 (1)
      mean: 113.405 (0.444726)
      standard deviation: 26.3364 (0.10328)
      kurtosis: 1.42366
      skewness: -1.15852
      entropy: 0.819023
  Image statistics:
    Overall:
      min: 0  (0)
      max: 255 (1)
      mean: 142.226 (0.557749)
      standard deviation: 30.1145 (0.118096)
      kurtosis: -0.166765
      skewness: -0.687147
      entropy: 0.804755
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 3283x4714+0+0
  Dispose: Undefined
  Iterations: 0
  Scene: 1
  Compression: LZW
  Orientation: TopLeft
  Convex hull: 0,0 3282,0 3282,4712 3281,4713 0,4713 0,0
  Minimum bounding box: 3282,0 3282,4713 0,4713 0,0
  Properties:
    date:create: 2023-11-30T11:00:56+00:00
    date:modify: 2023-11-28T08:03:48+00:00
    minimum-bounding-box:_p: 0,0
    minimum-bounding-box:_q: 3282,0
    minimum-bounding-box:_v: 3281,4713
    minimum-bounding-box:angle: 0
    minimum-bounding-box:area: 1.54681e+07
    minimum-bounding-box:height: 3282
    minimum-bounding-box:unrotate: -0
    minimum-bounding-box:width: 4713
    signature: 53da4cc911ee4652186674d764014c7e55d72add656c17e40e9b6e9a532373e1
    tiff:alpha: unspecified
    tiff:endian: lsb
    tiff:photometric: RGB
    tiff:rows-per-strip: 53
  Tainted: False
  Filesize: 33.2461MiB
  Number pixels: 15.4761M
  Pixels per second: 6.3228MP
```

**Comment by snibgo** [comment]
I don't know what the problem is.

Try it without `-tile 3x2 -geometry +10+20 -frame 10 -pointsize 60 -title "test"`. If that fixes the problem, then replace each option one by one to isolate the problem.

But if the problem still occurs without those options, then try inserting `-strip` and `+repage` before writing the output.

Can you zip your input files and upload the zip file?

**Reply by ferropasha** [reply]
I rewrote the command like this
`magick montage ( "h:\test\1.tif" -set label "1" ) ( "h:\test\2.tif" -set label "2" ) ( "h:\test\3.tif" -set label "3" ) ( "h:\test\4.tif" -set label "4" ) ( "h:\test\5.tif" -set label "5" ) ( "h:\test\6.tif" -set label "6" )  test_montage.jpg`
No dice. Adding -strp +repage before output file doesn't change anything as well. 
`magick montage ( "h:\test\1.tif" -set label "1" ) ( "h:\test\2.tif" -set label "2" ) ( "h:\test\3.tif" -set label "3" ) ( "h:\test\4.tif" -set label "4" ) ( "h:\test\5.tif" -set label "5" ) ( "h:\test\6.tif" -set label "6" )  -strip +repage test_montage.jpg`
Neither does separately rewriting every image before feeding it into montage like this
`magick "h:\test\1.tif" -strip +repage "h:\test\1.tif" `
I put the sample files (the ones before -strip +repage applied) and the last failed montage output file in the zip archive. The problem is it is too big, 212 mb, I can't attach it. Will you accept a google drive link?
PS I also run one more experiment with the files. Apparently no matter how many files I take the result is the same even if I leave just files 1.tif and 2.tif in the command montage still swaps them around.

**Comment by snibgo** [comment]
> Will you accept a google drive link?

Yes, but not that size. 212 mb is too large for my lousy internet connection. Can you reproduce the problem with just 2 or 3 inputs?

Are all your inputs the same size?

Another possibility is to read the inputs with one `magick` command (not `magick montage`) and pipe the results in MIFF format to a `magick montage` command. Like this, but using your files instead of `rose:` and `wizard:`:
```
magick ( rose: -set label "my rose" ) ( wizard: -set label "my wizard" ) +repage -strip miff:- | magick montage miff:- -tile 3x2 -geometry +10+20 -frame 10 -pointsize 60 -title "test" x.png
```

**Reply by ferropasha** [reply]
Each file is 30+ mb. They are scans of first pages from old newspapers so expected to be rather large. Yes, the problem is still there even if  if I leave just files 1.tif and 2.tif in the command, montage swaps them around anyway. Will 2 files be good?

**Comment by snibgo** [comment]
> Will 2 files be good?

Yes.

**Reply by ferropasha** [reply]
It may not be necessary because you are a true sensei and master of the craft, Snibgo.  Your last idea with intermediate miff finally did the trick. Great thanks. I salute you.
Though I can still upload the archive and post the link if you are interested in reproducing the problem for yourself.

**Comment by snibgo** [comment]
I'm glad the problem is fixed.

> I can still upload the archive and post the link if you are interested in reproducing the problem for yourself.

No, thanks.
